tangelo/algorithms/variational/vqe_solver.py

```diff
@@ -213,43 +208,20 @@ def build(self):
-                        self.ansatz_options["reference_state"] = "zero"
-                    self.ansatz = HEA(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                elif self.ansatz == BuiltInAnsatze.UpCCGSD:
-                    self.ansatz = UpCCGSD(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                    self.ansatz.default_reference_state = "HF" if self.ref_state is None else "zero"
-                elif self.ansatz == BuiltInAnsatze.QMF:
-                    self.ansatz = QMF(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                elif self.ansatz == BuiltInAnsatze.QCC:
-                    self.ansatz = QCC(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                elif self.ansatz == BuiltInAnsatze.VSQS:
-                    self.ansatz = VSQS(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                elif self.ansatz == BuiltInAnsatze.UCCGD:
-                    self.ansatz = UCCGD(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-                    self.ansatz.default_reference_state = "HF" if self.ref_state is None else "zero"
-                elif self.ansatz == BuiltInAnsatze.ILC:
-                    self.ansatz = ILC(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
+                if self.ansatz in {BuiltInAnsatze.UCC1, BuiltInAnsatze.UCC3}:
+                    n_var_params = 1 if self.ansatz == BuiltInAnsatze.UCC1 else 3
+                    self.ansatz = agen.RUCC(n_var_params)
+                elif self.ansatz in self.builtin_ansatze:
+                    self.ansatz = eval(f"agen.{self.ansatz.name}(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)")
                 else:
                     raise ValueError(f"Unsupported ansatz. Built-in ansatze:\n\t{self.builtin_ansatze}")
-            elif not isinstance(self.ansatz, Ansatz):
+            elif not isinstance(self.ansatz, agen.Ansatz):
                 raise TypeError(f"Invalid ansatz dataype. Expecting instance of Ansatz class, or one of built-in options:\n\t{self.builtin_ansatze}")
 
         # Building with a qubit Hamiltonian.
-        elif self.ansatz in [BuiltInAnsatze.HEA, BuiltInAnsatze.VSQS]:
-            if self.ansatz == BuiltInAnsatze.HEA:
-                self.ansatz = HEA(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-            elif self.ansatz == BuiltInAnsatze.VSQS:
-                self.ansatz = VSQS(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)
-        elif not isinstance(self.ansatz, Ansatz):
+        elif self.ansatz in {BuiltInAnsatze.HEA, BuiltInAnsatze.VSQS}:
+            self.ansatz = eval(f"agen.{self.ansatz.name}(self.molecule, self.qubit_mapping, self.up_then_down, **self.ansatz_options)")
```

Comment:
Pretty ugly way to do it.

I'd rather you have a dictionary `d` that maps the enum `BuiltInAnsatze` values to the name of the class you are instantiating. Then the class name is given by `d[BuiltInAnsatze.HEA]` and you can instantiate with `d[BuiltInAnsatze.HEA](...)`

If this dictionary is mostly trivially defined using the `name` attribute, so be it.

Reply:
The enum class itself is kind of a dictionary. Creating another one would be redundant according to me. Now we don't deal with the `int` in the `BuiltInAnsatze`, and we could change those values to the objects.
```python
class BuiltInAnsatze(Enum):
    UCCSD = agen.UCCSD
    UCC1 = agen.RUCC
    UCC3 = agen.RUCC
    HEA = agen.HEA
...
```
And then we could do (if `self.ansatz` is a `BuiltInAnsatze` -> `self.ansatz(self.molecule, ...)`. This was the other option I considered.